### PR TITLE
Data management improvements

### PR DESCRIPTION
**Pull request by hweihwang**

### CODE REVIEW

File: websocket_server/BackupManager.js [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+	async writeBackupFile(roomId, backupData) {
+		const backupFile = path.join(
+			this.backupDir,
+			`${roomId}_${backupData.timestamp}.bak`,
+		)
+		const tempFile = `${backupFile}.tmp`
+
+		const compressed = await gzip(JSON.stringify(backupData))
+		await fs.writeFile(tempFile, compressed)
+		await fs.rename(tempFile, backupFile)
```
Comment (grnd-alt) [inline websocket_server/BackupManager.js:197]: What's the advantage of writing the tempFile to disk if we're instantly renaming it?

**Review by grnd-alt — COMMENTED** [review]
🎉 code quality is good, and changes would reduce reported errors, great work.

I would think having file operations for each websocket sync will increase performance loads significantly. have you thought about doing the backups only every X amount of syncs?

File: websocket_server/BackupManager.js (hunk 1 above)
Comment (hweihwang) [reply websocket_server/BackupManager.js:197]: Good points, this was what I searched at that time, not really tested through too :)


**Temporary File Usage**

The temporary file pattern (writing to .tmp then renaming) is actually a common safety practice for file operations, even if the rename happens immediately after. Here's why:
- **Atomic Operations**: The rename operation is atomic on most filesystems, while writeFile is not. This means that if the process crashes during a write, you won't end up with a partially written backup file.
- **Preventing Corruption**: If another process is reading the backup file, the temporary file approach ensures they always get a complete file.
- **Filesystem Caching**: The actual disk write might be buffered by the OS, so the rename provides a guarantee that the write is complete.

**Comment by hweihwang** [comment]
@grnd-alt Really valid points on server/files operations there, I thought about that too, the current implementation had some tweaks for that:
+ We had a file locking mechanism with a default timeout of 5 seconds (this.lockTimeout = options.lockTimeout || 5000) to prevent too frequent writes
+ The backup operations are called asynchronously and not really block the main thread:
```js
if (updatedData) {
    await this.updateRoomWithData(room, updatedData, users, lastEditedUser)
    this.createRoomBackup(room.id, room) // async call without await
}
```
=> Would be better if can move it to a separate process (a queue processor) to process saving since it's not really critical

+ I currently think of incorporating with some client side storage (IndexedDB/LocalStorage) to reduce some server loads/operations and prevent data loss too, since one user might only access few rooms at a time compared to server with lots amount of data there. Let me know your thoughts on this one @grnd-alt , thank you!

**Comment by grnd-alt** [comment]
@hweihwang Yes storing data in localstorage is a good idea, I think we should investigate it later on, as it might get complex to have divergent versions from multiple users, that need to be reconciled when restoring from the browser storage.

**Review by grnd-alt — APPROVED** [review]
🎉 as said before the code is great.
IndexDB to be discussed later

**Comment by juliusknorr** [comment]
Anything still pending for the merge?

**Comment by hweihwang** [comment]
I'm still thinking a bit about the approach since the board can still work isolatedly without much needed from nextcloud server
=> So user might no longer needed to correct their configurations to connect from the whiteboard server to nextcloud server
=> Data will be stay in the node server and never go to nextcloud storage
=> Might introduce some security issues about read and write permissions to the file
=> ...

Not sure they're 100% related or might be in another ticket. Will think a bit more about that. Want to know your thoughts on this @juliusknorr @grnd-alt too! Thank you!

**Pull request merged**